app/models/route_stop_pattern.rb

```diff
@@ -0,0 +1,230 @@
+            # other stop        * current stop *|-------- other stop ---------------|
+            total_distance += stop[:geometry].distance(previous_stop[:geometry])
+            distances << total_distance
+          end
+        else
+          # current stop is within the line geometry, or the first stop past the last
+          # endpoint of the line geometry.
+          #                   |      <-- line geometry -->           |
+          # other stop        |-------- * current stop * ------------|     other stop
+          #                                   OR
+          # other stop        |-------- other stop ------------------| * current stop *     other stop
+          total_distance += RGeo::Feature.cast(splits[0], RouteStopPattern::GEOFACTORY).length
+          if splits[1].nil?
+            # current stop is the first past the last geometry endpoint, so
+            # we need to tack on that extra distance between that stop and the
+            # stop before, if such a space exists.
+            if !geometry_endpoint_reached
+              total_distance += stop[:geometry].distance(cast_route.end_point)
+            end
+            geometry_endpoint_reached = true
+          else
+            cast_route = splits[1]
+          end
+
+          # tack on the extra space between the previous stop and the first point
+          # of the line geometry, if any. This is added to the distance measurement
+          # of the previous stop.
+          if !geometry_start_reached
+            distances[i-1] += previous_stop[:geometry].distance(splits[0].start_point) if i!=0
+          end
+          geometry_start_reached = true
+          distances << total_distance
+        end
+      end
+    end
+    distances
+  end
+
+  def evaluate_geometry(trip, stop_points)
+    # makes judgements on geometry so modifications can be made by tl_geometry
+    issues = []
+    if trip.shape_id.nil? || self.geometry[:coordinates].empty?
+      issues << :empty
+    end
+    # more evaluations can go here. e.g. has outlier stop
+    return (issues.size > 0), issues
+  end
+
+  def tl_geometry(stop_points, issues)
+    # modify rsp geometry based on issues hash from evaluate_geometry
+    if issues.include?(:empty)
+      # create a new geometry from the trip stop points
+      self.geometry = RouteStopPattern.line_string(stop_points)
+      self.is_generated = true
+      self.is_modified = true
+    end
+    # more geometry modification can go here
+  end
+
+  scope :with_trips, -> (search_string) { where{trips.within(search_string)} }
+  scope :with_stops, -> (search_string) { where{stop_pattern.within(search_string)} }
+
+  ##### FromGTFS ####
+  include FromGTFS
+  def self.from_gtfs(trip, stop_pattern, shape_points)
+    raise ArgumentError.new('Need at least two stops') if stop_pattern.length < 2
+    rsp = RouteStopPattern.new(
+      stop_pattern: stop_pattern,
+      geometry: self.line_string(shape_points.uniq)
```

Comment:
perhaps useful: http://stackoverflow.com/questions/2516312/remove-adjacent-identical-elements-in-a-ruby-array